### PR TITLE
A0-569: connection types

### PR DESCRIPTION
**Pull request by pmikolajczyk41**
# Description

This PR introduces two new types of connection: `SignedConnection` and `RootConnection`. This improves the API of `aleph-client` by moving requirements for e.g. signer presence to a type level instead of docstring or runtime check.

## Type of change

Please delete options that are not relevant.

- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

# Checklist:

- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Infrastructure updated accordingly
- [x] I have made corresponding changes to the documentation
- [ ] New documentation created
- [ ] Bump `spec_version` and `transaction_version` if relevant
- [x] Bump `aleph-client` version if relevant

### CODE REVIEW

File: e2e-tests/src/test/validators_change.rs [hunk 1]
```diff
@@ -18,9 +19,10 @@ pub fn change_validators(config: &Config) -> anyhow::Result<()> {
     let mut accounts = accounts_from_seeds(seeds);
     let sudo = get_sudo(config);
 
-    let connection = create_connection(node).set_signer(sudo);
+    let connection = RootConnection::new(SignedConnection::new(create_connection(node), sudo));
```
Comment (fbielejec) [inline e2e-tests/src/test/validators_change.rs:22]: this API is not super ergonomic :/ sucks there's no method overloading in Rust, but maybe `new` could take url and sudo keypair, and then have `from_signed_connection` / `from_any_connection` for "type cast constructors" WYT?

File: aleph-client/src/lib.rs [hunk 2]
```diff
@@ -71,6 +71,97 @@ pub type Header = GenericHeader<BlockNumber, BlakeTwo256>;
 pub type KeyPair = sr25519::Pair;
 pub type Connection = Api<KeyPair, WsRpcClient>;
 
+/// 'Castability' to `Connection`.
+///
+/// Direct casting is often more handy than generic `.into()`. Justification: `Connection` objects
+/// are often passed to some macro like `compose_extrinsic!` and thus there is not enough
+/// information for type inferring required for `Into<Connection>`.
+pub trait AnyConnection: Clone {
+    fn as_con(&self) -> Connection;
```
Comment (fbielejec) [inline aleph-client/src/lib.rs:80]: can we call it without abbreviations, `as_connection`

File: aleph-client/src/staking.rs [hunk 3]
```diff
@@ -183,29 +198,33 @@ pub fn batch_bond(
...
-    let xt = connection.staking_nominate(vec![GenericAddress::Id(nominee_account_id)]);
+    let xt = connection
+        .as_con()
+        .staking_nominate(vec![GenericAddress::Id(nominee_account_id)]);
     send_xt(connection, xt, Some("nominate"), XtStatus::InBlock);
 }
 
 pub fn batch_nominate(
-    connection: &Connection,
+    connection: &RootConnection,
```
Comment (Marcin-Radecki) [inline aleph-client/src/staking.rs:211]: We don't need to be sudo afaik

File: bin/cliain/src/keys.rs [hunk 4]
```diff
@@ -1,31 +1,31 @@
...
 };
 use log::info;
 use primitives::staking::MIN_VALIDATOR_BOND;
 use substrate_api_client::{AccountId, XtStatus};
 
-pub fn prepare_keys(connection: Connection, controller_account_id: AccountId) {
+pub fn prepare_keys(connection: RootConnection, controller_account_id: AccountId) {
     staking_bond(
-        &connection,
+        &connection.as_signed(),
```
Comment (Marcin-Radecki) [inline bin/cliain/src/keys.rs:11]: I understand now why `staking_bond` takes `RootConnection` , can we somehow cast `RootConnection ` to signed connections here?

File: bin/cliain/src/keys.rs [hunk 5]
```diff
@@ -1,31 +1,31 @@
...
+pub fn set_keys(connection: SignedConnection, new_keys: String) {
     set(
         &connection,
         SessionKeys::try_from(new_keys).expect("Failed to parse keys"),
         XtStatus::InBlock,
     );
 }
 
-pub fn rotate_keys(connection: Connection) {
+pub fn rotate_keys(connection: RootConnection) {
```
Comment (Marcin-Radecki) [inline bin/cliain/src/keys.rs:28]: We don't need to be sudo, rotating keys is done via RPC call anyway.

**Review by Marcin-Radecki — CHANGES_REQUESTED** [review]
Just a few comments about `RootConnection` not being required in pallet utility wrapper calls. Otherwise, decent rework and value-added.

File: aleph-client/src/staking.rs [hunk 6]
```diff
@@ -147,25 +160,27 @@ pub fn payout_stakers_and_assert_locked_balance(
 }
 
 pub fn batch_bond(
-    connection: &Connection,
+    connection: &RootConnection,
```
[the remark this answers is not in the document]
Comment (pmikolajczyk41) [reply aleph-client/src/staking.rs:163]: that is true, but in this method we are performing bonding with `sudo_as`

File: aleph-client/src/staking.rs (hunk 3 above)
Comment (pmikolajczyk41) [reply aleph-client/src/staking.rs:211]: the same thing as with `batch_bond`

Comment (Marcin-Radecki) [reply aleph-client/src/staking.rs:211]: Right, I missed that

File: aleph-client/src/staking.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (Marcin-Radecki) [reply aleph-client/src/staking.rs:163]: Right

File: bin/cliain/src/keys.rs (hunk 4 above)
Comment (Marcin-Radecki) [reply bin/cliain/src/keys.rs:11]: Nvm I was under the impression this code won't compile but it will.

File: aleph-client/src/lib.rs (hunk 2 above)
Comment (pmikolajczyk41) [reply aleph-client/src/lib.rs:80]: renamed

File: bin/cliain/src/keys.rs (hunk 5 above)
Comment (pmikolajczyk41) [reply bin/cliain/src/keys.rs]: thank you, changed

File: e2e-tests/src/test/validators_change.rs (hunk 1 above)
Comment (pmikolajczyk41) [reply e2e-tests/src/test/validators_change.rs:22]: changed somehow - see whether it works better now

**Review by fbielejec — APPROVED** [review]

**Review by Marcin-Radecki — APPROVED** [review]

**Pull request merged**